homeassistant/components/medcom_ble/config_flow.py

```diff
@@ -0,0 +1,160 @@
+            raise AbortFlow(
+                "cannot_connect",
+                description_placeholders={
+                    "error connecting to": discovery_info.address
+                },
+            )
+
+        inspector = MedcomBleDeviceData(_LOGGER)
+
+        return await inspector.update_device(ble_device)
+
+    async def async_step_bluetooth(
+        self, discovery_info: BluetoothServiceInfo
+    ) -> FlowResult:
+        """Handle the bluetooth discovery step."""
+        _LOGGER.debug("Discovered BLE device: %s", discovery_info.name)
+        await self.async_set_unique_id(discovery_info.address)
+        self._abort_if_unique_id_configured()
+
+        self._discovery_info = discovery_info
+        self.context["title_placeholders"] = {
+            "name": human_readable_name(
+                None, discovery_info.name, discovery_info.address
+            )
+        }
+
+        return await self.async_step_bluetooth_confirm()
+
+    async def async_step_bluetooth_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Confirm discovery."""
+        if user_input is not None:
+            return self.async_create_entry(
+                title=self.context["title_placeholders"]["name"], data={}
+            )
+
+        self._set_confirm_only()
+        return self.async_show_form(
+            step_id="bluetooth_confirm",
+            description_placeholders=self.context["title_placeholders"],
+        )
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the user step to pick discovered device."""
+        if user_input is not None:
+            address = user_input[CONF_ADDRESS]
+            await self.async_set_unique_id(address, raise_on_progress=False)
+            self._abort_if_unique_id_configured()
+            discovery = self._discovered_devices[address]
+
+            self.context["title_placeholders"] = {
+                "name": discovery.name,
+            }
+
+            self._discovered_device = discovery
+
+            return self.async_create_entry(title=discovery.name, data={})
+
+        current_addresses = self._async_current_ids()
+        for discovery_info in async_discovered_service_info(self.hass):
+            address = discovery_info.address
+            if address in current_addresses or address in self._discovered_devices:
+                continue
+
+            try:
+                device = await self._get_device_data(discovery_info)
+            except (BleakError, AbortFlow):
+                return self.async_abort(reason="cannot_connect")
+            except Exception as err:  # pylint: disable=broad-except
+                _LOGGER.exception(
+                    "Error occurred reading information from %s: %s",
+                    discovery_info.address,
+                    err,
+                )
+                return self.async_abort(reason="unknown")
```

Comment:
We should only connect once they have selected the device they want to set up, we should move this to the next step.

Since you have a very unique service uuid to match on, only show them devices that match that service uuid.

Once they have selected, the device, move to a new step and verify the connection.  You should also do this from `async_step_bluetooth` as well instead of confirming right away since we don't actually know if we can connect

Reply:
I think I understand what you mean. What's the normal sequence of steps in this config_flow.py ? Any example of other integrations doing this properly?  between `async_step_bluetooth`, `async_step_bluetooth_confirm`, `async_step_user` etc, it's not always super obvious what the flow looks like.

Reply:
On the service uuid match: since it's specified in the manifest, won't the integration already have filtered devices that match this service uuid? Does it have to be checked in the config flow as well?

Reply:
> On the service uuid match: since it's specified in the manifest, won't the integration already have filtered devices that match this service uuid? Does it have to be checked in the config flow as well?

It will to reach the bluetooth step, but not the user step

Reply:
> I think I understand what you mean. What's the normal sequence of steps in this config_flow.py ? Any example of other integrations doing this properly? between `async_step_bluetooth`, `async_step_bluetooth_confirm`, `async_step_user` etc, it's not always super obvious what the flow looks like.

A normal success path looks like:

async_step_user (chooses device) -> async_step_confirm -> complete
async_step_bluetooth (from discovery) -> async_step_confirm -> complete


The confirm step might be the same or you might have a different step to confirm depending on how much info you have.  I think you can use the same in this case

Reply:
Gotcha... Actually, is the `async_step_user` function really required? it would only be called to manually add a device, but in practice, those devices will always be autodiscovered, unless there is a use-case I'm not aware of? I have updated the code to split the Bluetooth confirm step and first device connection attempt as suggested, but while doing that I realized I never use the `async_step_user` function at all...

Reply:
If the user closes the discovery there is no way to get it back so they have to go though the user flow at that point

Reply:
(actually it looks not that bad to configure properly, even if I'm not entirely sure why it's useful)

Reply:
Just need to update the tests now :)